lua/luasnip/extras/fmt.lua

```diff
@@ -0,0 +1,159 @@
+            prev_end = end_
+            -- interpret the placeholder
+            local placeholder = fmt:sub(start + 1, end_ - 1)
+            local is_num = placeholder:match('^%d+$')
+            local key
+            if is_num then  -- numbered placeholder
+                last_index = tonumber(placeholder)
+                key = last_index
+            elseif placeholder == '' then  -- auto-numbered placeholder
+                key = last_index + 1
+                last_index = key
+            else  -- named placeholder
+                key = placeholder
+            end
+            add_arg(key)
+        end
+        -- end
+    end
+
+    return elements
+end
+
+-- Find the largest common indent in a list of lines and remove it.
+local function remove_common_indent(lines)
+    local max_indent = 100000
+    for _, line in ipairs(lines) do
+        -- ignore lines with whitespace only
+        local match = line:match('^(%s*)%S')
+        if match then
+            max_indent = math.min(max_indent, #match)
+        end
+    end
+    if max_indent > 0 then
+        lines = vim.tbl_map(function(line)
+            return line:sub(max_indent + 1)
+        end, lines)
+    end
+    return lines
+end
+
+-- Use a format string to replace placeholders with nodes.
+-- The format is similar to Python formatting, e.g. `Some {} then {3} and {name}.`
+-- See `replace_placeholders` documentation for details.
+--
+-- Arguments:
+--   str: format string
+--   nodes: snippet node or list of nodes
+--   opts: optional table
+--     trim_empty: boolean, remove whitespace-only first/last lines, default true
+--     dedent: boolean, remove all common indent in `str`, default true
+-- Returns: list of snippet nodes
+local function format_nodes(str, nodes, opts)
+    opts = opts or {}
+    opts.trim_empty = opts.trim_empty ~= nil and opts.trim_empty or true
+    opts.dedent = opts.dedent ~= nil and opts.dedent or true
+
+    -- allow to pass a single node
+    if type(nodes.is_node) == 'function' and nodes.is_node() then
+        nodes = {nodes}
+    end
+
+    -- optimization: avoid splitting multiple times
+    local lines = nil
+
+    -- this allows to use [[...]] strings ignoring the first and last lines
+    if opts.trim_empty then
+        lines = vim.split(str, '\n', true)
+        if lines[1]:match('^%s*$') then
+            table.remove(lines, 1)
+        end
+        if lines[#lines]:match('^%s*$') then
+            table.remove(lines)
+        end
+    end
+
+    -- remove common indent
+    if opts.dedent then
+        lines = lines or vim.split(str, '\n', true)
+        lines = remove_common_indent(lines)
+    end
+
+    if lines then
+        str = table.concat(lines, '\n')
+    end
+
+    local parts = replace_placeholders(str, nodes)
+    return vim.tbl_map(function(part)
+        -- wrap strings in text nodes
+        if type(part) == 'string' then
+            return ls.text_node(vim.split(part, '\n', true))
```

Comment:
Afaict luasnip is only `require`d to get access to `textNodes`, right? It'd be nicer to directly `require("luasnip.nodes.textNode")`

Reply:
Sure